src/vendored/CowWrapper.sol

```diff
@@ -0,0 +1,178 @@
+    function wrappedSettle(
+        bytes calldata settleData,
+        bytes calldata wrapperData
+    ) external {
+        // Revert if not a valid solver
+        require(AUTHENTICATOR.isSolver(msg.sender), NotASolver(msg.sender));
+
+        // Require wrapper data to contain at least the next settlement address (20 bytes)
+        require(wrapperData.length >= 20, WrapperHasNoSettleTarget(wrapperData.length, 20));
+
+        // Delegate to the wrapper's custom logic
+        _wrap(settleData, wrapperData);
+    }
+
+    /// @notice Parses and validates wrapper-specific data
+    /// @dev Must be implemented by concrete wrapper contracts. Used for pre-execution validation.
+    ///      The implementation should consume its wrapper-specific data and return the remainder.
+    /// @param wrapperData The full wrapper data to parse
+    /// @return remainingWrapperData The portion of wrapper data not consumed by this wrapper
+    function parseWrapperData(bytes calldata wrapperData) external virtual view returns (bytes calldata remainingWrapperData);
+
+    /// @notice Internal function containing the wrapper's custom logic
+    /// @dev Must be implemented by concrete wrapper contracts. Should execute custom logic
+    ///      then eventually call _internalSettle() to continue the settlement chain.
+    /// @param settleData ABI-encoded call to CowSettlement.settle()
+    /// @param wrapperData The wrapper data, which may be parsed and consumed as needed
+    function _wrap(bytes calldata settleData, bytes calldata wrapperData) internal virtual;
+
+    /// @notice Continues the settlement chain by calling the next wrapper or settlement contract
+    /// @dev Extracts the next target address from wrapperData and either:
+    ///      - Calls CowSettlement.settle() directly if no more wrappers remain, or
+    ///      - Calls the next CowWrapper.wrappedSettle() to continue the chain
+    /// @param settleData ABI-encoded call to CowSettlement.settle()
+    /// @param wrapperData Remaining wrapper data starting with the next target address (20 bytes)
+    function _internalSettle(bytes calldata settleData, bytes calldata wrapperData) internal {
+        // Extract the next settlement address from the first 20 bytes of wrapperData
+        address nextSettlement = address(bytes20(wrapperData[:20]));
+
+        // Skip past the address we just read
+        wrapperData = wrapperData[20:];
+
+        if (wrapperData.length == 0) {
+            // No more wrapper data - we're calling the final settlement contract
+            // Verify the settle data has the correct function selector
+            require(bytes4(settleData[:4]) == CowSettlement.settle.selector, InvalidSettleData(settleData));
+
+            // Call the settlement contract directly with the settle data
+            (bool success, bytes memory returnData) = nextSettlement.call(settleData);
```

Comment:
Although beautiful to take this from the last parameter. Given there's only one settlement, we can also assume once there's no more bytes its time to call the settlement, whose address is hardcoded by constructor param